File: src/homemade/game/model/combo/ComboEffectVendor.java
package homemade.game.model.combo;

import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedList;

public class ComboEffectVendor {

    public void addComboEffectsForTier(LinkedList<ComboEffect> effects, int tier) {
        ComboEffect[] pricedEffects = ComboEffect.values();
        Comparator<ComboEffect> comparator = ComboEffect.getComparator();

        Arrays.sort(pricedEffects, comparator);

        for (int i = 0; i < pricedEffects.length && tier > 0; i++) {
            ComboEffect effect = pricedEffects[i];
            int price = effect.price;

            int given = tier / price;
            tier -= given * price;

            for (int j = 0; j < given; j++) {
                effects.add(effect);
            }
        }

        effects.sort(comparator);
    }
}

File: game-loop/src/main/java/homemade/game/pipeline/operations/ComboEffectVendor.kt
package homemade.game.pipeline.operations

import homemade.game.model.combo.ComboEffect

class ComboEffectVendor {
    fun addComboEffectsForTier(effects: ArrayList<ComboEffect>, tier: Int) {
        var sumToSpend = tier
        val pricedEffects = ComboEffect.values().sortedBy { -it.price }
        var i = 0
        while (i < pricedEffects.size && sumToSpend > 0) {
            val effect = pricedEffects[i]
            val price = effect.price
            val given = sumToSpend / price
            sumToSpend -= given * price
            for (j in 0 until given) {
                effects.add(effect)
            }
            i++
        }
        effects.sortBy { -it.price }
    }
}
